controllers/envoysecuritypolicy_referencegrant_controller.go

```diff
@@ -0,0 +1,160 @@
+package controllers
+
+import (
+	"context"
+	"encoding/json"
+
+	egv1alpha1 "github.com/envoyproxy/gateway/api/v1alpha1"
+	"github.com/go-logr/logr"
+	kuadrantv1beta1 "github.com/kuadrant/kuadrant-operator/api/v1beta1"
+	kuadrantenvoygateway "github.com/kuadrant/kuadrant-operator/pkg/envoygateway"
+	kuadrantgatewayapi "github.com/kuadrant/kuadrant-operator/pkg/library/gatewayapi"
+	"github.com/kuadrant/kuadrant-operator/pkg/library/kuadrant"
+	"github.com/kuadrant/kuadrant-operator/pkg/library/mappers"
+	"github.com/kuadrant/kuadrant-operator/pkg/library/reconcilers"
+	"github.com/kuadrant/kuadrant-operator/pkg/library/utils"
+	"github.com/samber/lo"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/utils/ptr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	gatewayapiv1 "sigs.k8s.io/gateway-api/apis/v1"
+	gatewayapiv1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
+)
+
+const (
+	KuadrantReferenceGrantName = "kuadrant-authorization-rg"
+)
+
+// EnvoySecurityPolicyReferenceGrantReconciler reconciles ReferenceGrant objects for auth
+type EnvoySecurityPolicyReferenceGrantReconciler struct {
+	*reconcilers.BaseReconciler
+}
+
+//+kubebuilder:rbac:groups=gateway.networking.k8s.io,resources=referencegrants,verbs=get;list;watch;create;update;patch;delete
+
+func (r *EnvoySecurityPolicyReferenceGrantReconciler) Reconcile(eventCtx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := r.Logger().WithValues("Kuadrant", req.NamespacedName)
+	logger.Info("Reconciling SecurityPolicy ReferenceGrant")
+	ctx := logr.NewContext(eventCtx, logger)
+
+	kObj := &kuadrantv1beta1.Kuadrant{}
+	if err := r.Client().Get(ctx, req.NamespacedName, kObj); err != nil {
+		if apierrors.IsNotFound(err) {
+			logger.Info("no kuadrant object found")
+			return ctrl.Result{}, nil
+		}
+		logger.Error(err, "failed to get kuadrant object")
+		return ctrl.Result{}, err
+	}
+
+	if logger.V(1).Enabled() {
+		jsonData, err := json.MarshalIndent(kObj, "", "  ")
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		logger.V(1).Info(string(jsonData))
+	}
+
+	rg, err := r.securityPolicyReferenceGrant(ctx, kObj.Namespace)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if err := r.ReconcileResource(ctx, &gatewayapiv1beta1.ReferenceGrant{}, rg, kuadrantenvoygateway.SecurityPolicyReferenceGrantMutator); err != nil && !apierrors.IsAlreadyExists(err) {
```

Comment:
Maybe catch conflic error. But "alreadyExists" error can happen?

Reply:
I think this only ignores the error from where there's a race between the `Get` of the reference grant returning no reference grant, and it being created in the cluster before we call `Create`? But I guess in this case maybe we should error and try again

Reply:
But there is no race, is it? The only controller creating this refgrant is this very controller. So, if it does not exist, it will create, and if it exists, call the mutator to update if needed. I do not see how "alreadyExists" error might happen. I might be missing something, though.